backend/queues/listingCreatedProcessor.js

```diff
@@ -0,0 +1,181 @@
+      marketplaceAbi,
+      provider
+    )
+
+    // Load the transaction from the DB.
+    const transaction = await Transaction.findOne({
+      networkId: network.id,
+      shopId,
+      fromAddress,
+      hash: txHash
+    })
+    if (!transaction) {
+      throw new Error(
+        `No transaction found in the DB with hash ${txHash} for shop ${shopId} and wallet ${fromAddress}`
+      )
+    }
+
+    // Load the tx from the blockchain based on its hash.
+    queueLog(25, 'Loading tx from the blockchain')
+    const tx = await provider.getTransaction(txHash)
+    if (!tx) {
+      throw new Error(`Transaction with hash ${txHash} not found`)
+    }
+    log.info(`Loaded tx with hash ${txHash} from the network`)
+
+    // Wait for the tx to get mined.
+    queueLog(50, `Waiting for tx ${txHash} to get confirmed`)
+    log.info('Waiting for tx ${txHash} confirmation...')
+    confirmation = await _waitForListingCreatedTxConfirmation(marketplace, tx)
+    const { receipt, listingId } = confirmation
+
+    if (receipt.status) {
+      // Update the transaction in the DB.
+      const lid = new ListingID(listingId, network.networkId).toString() // fully qualified listingId.
+      await transaction.update({
+        status: TransactionStatuses.Confirmed,
+        blockNumber: receipt.blockNumber,
+        listingId: lid,
+        jobId
+      })
+      // Check the shop still does not have a listingId (a concurrent ListingCreated tx could
+      // have been mined while we were waiting for this tx to get mined).
+      await shop.load()
+      if (shop.listingId) {
+        log.info(
+          `ListingId ${shop.listingId} linked to shop ${shopId} while waiting for ${txHash}.`
+        )
+      } else {
+        await shop.update({ listingId: lid })
+        log.info(
+          `Shop ${shopId}: tx ${txHash} mined, updated listingId to ${lid}`
+        )
+      }
+    } else {
+      log.info(`Shop ${shopId}: tx with hash ${txHash} was reverted.`)
+      await transaction.update({
+        status: TransactionStatuses.Failed,
+        blockNumber: receipt.blockNumber,
+        jobId
+      })
+    }
+
+    queueLog(100, 'Finished')
+  } catch (e) {
+    // Log the exception and rethrow so that the job gets retried.
+    Sentry.captureException(e)
+    log.error(
+      `Waiting for ListingCreated for shop ${shopId} txHash ${txHash} failed:`,
+      e
+    )
+    throw e
+  }
+
+  return confirmation
+}
+
+/**
+ * Waits a marketplace listing creation transaction to get mined.
+ * Note: This is blocking. No timeout is set.
+ *
+ * @param {ethers.Contract} marketplace
+ * @param {ethers.Transaction} tx
+ * @returns {Promise<{receipt: ethers.TransactionReceipt, listingId: number|null, ipfsHash: string||null }>}
+ * @private
+ */
+async function _waitForListingCreatedTxConfirmation(marketplace, tx) {
+  // Wait for the tx to get mined with some blocks confirmation.
+  const receipt = await tx.wait(NUM_BLOCKS_CONFIRMATION)
```

Comment:
I'm wondering if we should check for outstanding transactions and confirmations manually each time there's a new block instead of relying on tx.wait? If someone speeds up their transaction and the txHash needs to be updated, might be a pain to clean up these tx.wait promises

Reply:
Hmm. Yeah good point about sped up transactions... For that case the approach of submitting the merchant's address and then watching for new events from that address would be more reliable. Let me look at that alternative approach. Thanks.